pkg/ccl/backupccl/restore_span_covering_test.go

```diff
@@ -1138,3 +1177,151 @@ func runTestRestoreEntryCover(t *testing.T, numBackups int, simpleImportSpans bo
+
+	type simpleRestoreSpanEntry struct {
+		span  roachpb.Span
+		paths []string
+	}
+
+	type testCase struct {
+		name                   string
+		requiredSpans          []roachpb.Span
+		backupSpans            []roachpb.Spans
+		expectedCover          []simpleRestoreSpanEntry
+		expectedCoverSimple    []simpleRestoreSpanEntry
+		expectedCoverGenerated []simpleRestoreSpanEntry
+	}
+
+	for _, tt := range []testCase{
+		{
+			name:          "file at start of span",
+			requiredSpans: []roachpb.Span{c.sp("a", "b")},
+			backupSpans: []roachpb.Spans{
+				{c.sp("a", "a")},
+			},
+			expectedCover: []simpleRestoreSpanEntry{
+				{span: c.sp("a", "b"), paths: []string{"1"}},
+			},
+		},
+		{
+			name:          "file at end of span",
+			requiredSpans: []roachpb.Span{c.sp("a", "b")},
+			backupSpans: []roachpb.Spans{
+				{c.sp("b", "b")},
+			},
+			expectedCover: []simpleRestoreSpanEntry{},
+		},
+		{
+			name:          "file at middle of span",
+			requiredSpans: []roachpb.Span{c.sp("a", "c")},
+			backupSpans: []roachpb.Spans{
+				{c.sp("b", "b")},
+			},
+			expectedCoverSimple: []simpleRestoreSpanEntry{
+				{span: c.sp("a", "c"), paths: []string{"1"}},
+			},
+			expectedCoverGenerated: []simpleRestoreSpanEntry{
+				{span: c.sp("b", "c"), paths: []string{"1"}},
+			},
+		},
+		{
+			name:          "sz0 file at end of prev file",
+			requiredSpans: []roachpb.Span{c.sp("a", "f")},
+			backupSpans: []roachpb.Spans{
+				{c.sp("a", "b"), c.sp("b", "b"), c.sp("b", "c")},
+			},
+			expectedCoverSimple: []simpleRestoreSpanEntry{
+				{span: c.sp("a", "b"), paths: []string{"1"}},
+				{span: c.sp("b", "f"), paths: []string{"1", "2", "3"}},
+			},
+			expectedCoverGenerated: []simpleRestoreSpanEntry{
+				{span: c.sp("a", "b"), paths: []string{"1"}},
+				{span: c.sp("b", "f"), paths: []string{"2", "3", "1"}},
+			},
+		},
+		{
+			name: "sz0 file contained by prev file",
+			requiredSpans: []roachpb.Span{
+				c.sp("a", "f"),
+			},
+			backupSpans: []roachpb.Spans{
+				{c.sp("a", "c"), c.sp("b", "b"), c.sp("b", "d")},
+			},
+			expectedCoverSimple: []simpleRestoreSpanEntry{
+				{span: c.sp("a", "c"), paths: []string{"1", "2", "3"}},
+				{span: c.sp("c", "f"), paths: []string{"1", "3"}},
+			},
+			expectedCoverGenerated: []simpleRestoreSpanEntry{
+				{span: c.sp("a", "b"), paths: []string{"1"}},
+				{span: c.sp("b", "f"), paths: []string{"2", "3", "1"}},
+			},
+		},
+		{
+			name: "sz0 file contained by following file",
+			requiredSpans: []roachpb.Span{
+				c.sp("a", "f"),
+			},
+			backupSpans: []roachpb.Spans{
+				{c.sp("b", "b"), c.sp("b", "c"), c.sp("b", "d")},
+			},
+			expectedCoverSimple: []simpleRestoreSpanEntry{
+				{span: c.sp("a", "b"), paths: []string{"1"}},
+				{span: c.sp("b", "c"), paths: []string{"1", "2", "3"}},
+				{span: c.sp("c", "f"), paths: []string{"2", "3"}},
+			},
+			expectedCoverGenerated: []simpleRestoreSpanEntry{
+				{span: c.sp("b", "f"), paths: []string{"1", "2", "3"}},
```

Comment:
this is an awesome test. 

 Some ways to increase coverage:
- an additional test case:
```
sz0 file contained by prev and following file

	backupSpans: []roachpb.Spans{
				{c.sp("a", "c"), c.sp("b", "b"), c.sp("b", "c"), c.sp("b", "d")},
			},
```

- i feel like there could be new cases where the sz0 file interacts with a cover generated by a previous backup layer. Suppose this test created up to n backup layers if there are n backup files and then randomly assigned backup files to those layers. I believe the makSimpleImportSpans covering should be the same, right? The ordering of the files for an entry may differ, but that doesn't matter.
- This change affects more than just zero sized file spans. Consider the adding the following test case:
```
fileSpan: {a,c)
requiredSpan: {c-d)
```
I was surprised that this case was not already in `TestRestoreEntryCoverExample`, but no need to change the test cases there. You could just add another test case in this test.